src/coreclr/src/vm/genanalysis.cpp

```diff
@@ -0,0 +1,79 @@
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#include "common.h"
+#include "genanalysis.h"
+
+GcGenAnalysisState gcGenAnalysisState = GcGenAnalysisState::Uninitialized;
+EventPipeSession* gcGenAnalysisEventPipeSession = nullptr;
+uint64_t gcGenAnalysisEventPipeSessionId = (uint64_t)-1;
+GcGenAnalysisState gcGenAnalysisConfigured = GcGenAnalysisState::Uninitialized;
+int64_t gcGenAnalysisGen = -1;
+int64_t gcGenAnalysisBytes = 0;
+int64_t gcGenAnalysisIndex = 0;
+
+/* static */ void GenAnalysis::Initialize()
+{
+#ifndef GEN_ANALYSIS_STRESS
+    if (gcGenAnalysisConfigured == GcGenAnalysisState::Uninitialized)
+    {
+        if (CLRConfig::IsConfigOptionSpecified(W("GCGenAnalysisGen")))
+        {
+            gcGenAnalysisGen = CLRConfig::GetConfigValue(CLRConfig::INTERNAL_GCGenAnalysisGen);
+            if (CLRConfig::IsConfigOptionSpecified(W("GCGenAnalysisBytes")))
+            {
+                gcGenAnalysisBytes = CLRConfig::GetConfigValue(CLRConfig::INTERNAL_GCGenAnalysisBytes);
+                if (CLRConfig::IsConfigOptionSpecified(W("GCGenAnalysisIndex")))
+                {
+                    gcGenAnalysisIndex = CLRConfig::GetConfigValue(CLRConfig::INTERNAL_GCGenAnalysisIndex);
+                    gcGenAnalysisConfigured = GcGenAnalysisState::Enabled;
+                }
+                else
+                {
+                    gcGenAnalysisConfigured = GcGenAnalysisState::Disabled;
+                }
+            }
+            else
+            {
+                gcGenAnalysisConfigured = GcGenAnalysisState::Disabled;
+            }
+        }
+    }
+    if ((gcGenAnalysisConfigured == GcGenAnalysisState::Enabled) && (gcGenAnalysisState == GcGenAnalysisState::Uninitialized))
+#endif
+    {
+        EnableGenerationalAwareSession();
+    }    
+}
+
+/* static */ void GenAnalysis::EnableGenerationalAwareSession()
+{
+    LPCWSTR outputPath = nullptr;
+    outputPath = GENAWARE_FILE_NAME;
+    NewHolder<EventPipeProviderConfiguration> pProviders = nullptr;
+    int providerCnt = 1;
+    pProviders = new EventPipeProviderConfiguration[providerCnt];
+    const uint64_t GCHeapAndTypeNamesKeyword        = 0x00001000000; // This keyword is necessary for the type names
+    const uint64_t GCHeapSurvivalAndMovementKeyword = 0x00000400000; // This keyword is necessary for the generation range data.
+    const uint64_t GCHeapDumpKeyword                = 0x00000100000; // This keyword is necessary for enabling walking the heap
+    const uint64_t TypeKeyword                      = 0x00000080000; // This keyword is necessary for enabling BulkType events
+    const uint64_t keyword                          = GCHeapAndTypeNamesKeyword|GCHeapSurvivalAndMovementKeyword|GCHeapDumpKeyword|TypeKeyword;
+    pProviders[0] = EventPipeProviderConfiguration(W("Microsoft-Windows-DotNETRuntime"), keyword, 5, nullptr);
+    gcGenAnalysisEventPipeSessionId = EventPipe::Enable(
+        outputPath,
+        1024,
```

Comment:
You could still use the COMPlus_EventPipeCircularMB here to make this configurable, right?

Reply:
I think this is a good suggestion and we should use the value specified by this env var if it's set. if you have a big heap and there are a lot of edges/etc 1gb may not be enough and since we already have an env var why not take advantage of it. @cshung, is there any reason why you wouldn't want to use the env var here?

Reply:
See https://github.com/dotnet/runtime/pull/40918